File: scripts/autofix/autofix.ts
#!/usr/bin/env bash
set -euo pipefail

CLEAN_ROOT="/mnt/24T/ABSTRACT_ENDEAVORS/scripts/RABBIT/aggregator_clean"
SOURCE_ROOT="/mnt/24T/ABSTRACT_ENDEAVORS/scripts/RABBIT/aggregator"
CMD=("npx" "tsx" "clients/main-cli.ts")

cd "$CLEAN_ROOT"

echo "▶ starting autofix runner in $CLEAN_ROOT"
echo

while true; do
  OUTPUT="$(mktemp)"

  if "${CMD[@]}" >"$OUTPUT" 2>&1; then
    echo "✅ program exited cleanly"
    rm -f "$OUTPUT"
    exit 0
  fi

  if ! grep -q "ERR_MODULE_NOT_FOUND" "$OUTPUT"; then
    cat "$OUTPUT"
    echo "❌ failed for non-module reason"
    exit 1
  fi

  MISSING="$(grep -oE "Cannot find module '([^']+)'" "$OUTPUT" | sed "s/.*'//;s/'//")"

  if [[ -z "$MISSING" ]]; then
    cat "$OUTPUT"
    echo "❌ could not extract missing module path"
    exit 1
  fi

  REL_PATH="${MISSING#$CLEAN_ROOT/}"
  SRC_FILE="$SOURCE_ROOT/$REL_PATH"
  DEST_FILE="$CLEAN_ROOT/$REL_PATH"

  echo "⚠️  missing: $REL_PATH"

  if [[ ! -f "$SRC_FILE" ]]; then
    echo "❌ source file not found either:"
    echo "   $SRC_FILE"
    exit 1
  fi

  mkdir -p "$(dirname "$DEST_FILE")"
  cp "$SRC_FILE" "$DEST_FILE"

  echo "✅ copied from source:"
  echo "   $SRC_FILE"
  echo "→  $DEST_FILE"
  echo
  rm -f "$OUTPUT"
done

File: scripts/autofix/autofix.py
#!/usr/bin/env python3
import subprocess, sys,shutil,os
ROOT_DIRECTORY = "/mnt/24T/ABSTRACT_ENDEAVORS/scripts/RABBIT"
MIRROR_BASENAME = "aggregator"
CLEAN_BASENAME = f"{MIRROR_BASENAME}_clean"
CLEAN_ROOT = os.path.join(ROOT_DIRECTORY,CLEAN_BASENAME)
ENTRY      = os.path.join(CLEAN_ROOT,"clients/main-cli.ts")
def copy_mirrored_file(output,txt):
    file_path = output.split(txt)[1].split("'")[0]
    dirname = os.path.dirname(file_path)
    basename = os.path.basename(file_path)
    filename,ext = os.path.splitext(basename)
    mirror_dirname = dirname.replace(CLEAN_BASENAME,MIRROR_BASENAME)
    if 'workflows/src' in str(mirror_dirname):
        mirror_dirname=mirror_dirname.replace('workflows/src','workflows/logData')
    input(mirror_dirname)
    for mirror_basename in os.listdir(mirror_dirname):
        mirror_filename,mirror_ext = os.path.splitext(mirror_basename)
        if mirror_filename == filename:
            mirror_file = os.path.join(mirror_dirname,mirror_basename)
            input(mirror_file)
            shutil.copy(mirror_file,dirname)
while True:
    result = subprocess.run(
        ["npx", "tsx", ENTRY],
        cwd=CLEAN_ROOT,
        capture_output=True,
        text=True,
    )

    output = (result.stdout + result.stderr).strip()
    
    if 'No such file or directory: ' in output:
        copy_mirrored_file(output,'No such file or directory: ')
        input(output)
    elif "Cannot find module '" in output:
        copy_mirrored_file(output,"Cannot find module '" )
        input(output)
    else:
        pretext = '/run/user/1000/gvfs/sftp:host=192.168.0.100,user=solcatcher'
        if 'file:///' in output:
            rel_path = output.split('file:///')[1].split(':')[0]
            nupath = f"{pretext}/{rel_path}"
            input(nupath)

        break
input(output)
sys.exit(result.returncode)
